File: scripts/algorithms/C/Copy List with Random Pointer/Copy List with Random Pointer.cpp
// Runtime: 20 ms (Top 20.49%) | Memory: 11.3 MB (Top 60.56%)
class Solution {
public:
    Node* copyRandomList(Node* head) {

        if(head==NULL){
            return NULL;
        }

        Node* temp= head;
        while(temp!=NULL){
            Node* m= new Node(temp->val);
            m->next= temp->next;
            temp->next= m;
            temp = temp->next->next;
        }
        temp= head;
        while(temp!=NULL){
            if(temp->random==NULL){
                temp->next->random=NULL;
            }
            else{
                temp->next->random= temp->random->next;
            }

            temp= temp->next->next;
        }
        temp= head;
        Node* p= new Node(-1);
        Node* r= p;
        Node* q= head;
        while(q!=NULL){
            q= q->next->next;
            p->next=temp->next;
            p=p->next;
            temp->next=q;
            temp= temp->next;
        }
        p->next=NULL;

        return r->next;
    }
};

File: scripts/algorithms/C/Copy List with Random Pointer/Copy List with Random Pointer.js
// Runtime: 68 ms (Top 90.10%) | Memory: 44 MB (Top 53.93%)
var copyRandomList = function(head) {
    if(!head) return head
    let curr = head;

    // manipulate the Linked List and store a copy node at every nod's next
    while(curr != null){
        let next = curr.next;
        curr.next = new Node(curr.val);
        curr.next.next = next
        curr = next
    }
    // console.log(head)

    // Setting random point of copied node according to original node
    curr = head;
    while(curr != null && curr.next != null){
        if(curr.random){
            curr.next.random = curr.random.next
        }else{
            curr.next.random = null
        }
        curr = curr.next.next
    }
    // console.log(head.next.next.next)

    // break the like made for linking copy node with original node
    curr = head;
    let ans = head.next;

    while(curr != null){
        let temp = curr.next;
        if(temp){
        curr.next = temp.next;
        curr = curr.next;
        }
        if(curr){
        temp.next = curr.next;
        temp = temp.next
        }
    }
    return ans
};